relayer/client.go

```diff
@@ -238,137 +234,182 @@ func CreateClient(ctx context.Context, src, dst *Chain, srcUpdateHeader, dstUpda
 	return true, nil
 }
 
-// UpdateClients updates clients for src on dst and dst on src given the configured paths
-func (c *Chain) UpdateClients(ctx context.Context, dst *Chain, memo string) (err error) {
-	var (
-		srcUpdateHeader, dstUpdateHeader ibcexported.Header
-		srch, dsth                       int64
-	)
-
-	if err = retry.Do(func() error {
-		srch, dsth, err = QueryLatestHeights(ctx, c, dst)
-		if err != nil {
-			return err
-		}
-		return nil
+// MsgUpdateClient queries for the current client state on dst,
+// then queries for the latest and trusted headers on src
+// in order to build a MsgUpdateClient message for dst.
+func MsgUpdateClient(
+	ctx context.Context,
+	src, dst *Chain,
+	srch, dsth int64,
+) (provider.RelayerMessage, error) {
+	var dstClientState ibcexported.ClientState
+	if err := retry.Do(func() error {
+		var err error
+		dstClientState, err = dst.ChainProvider.QueryClientState(ctx, dsth, dst.ClientID())
+		return err
 	}, retry.Context(ctx), RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
-		c.log.Info(
-			"Failed to get query latest heights when updating clients",
-			zap.String("src_chain_id", c.ChainID()),
-			zap.String("dst_chain_id", dst.ChainID()),
+		dst.log.Info(
+			"Failed to query client state when updating clients",
+			zap.String("client_id", dst.ClientID()),
 			zap.Uint("attempt", n+1),
 			zap.Uint("max_attempts", RtyAttNum),
 			zap.Error(err),
 		)
 	})); err != nil {
-		return err
+		return nil, err
 	}
 
-	if err = retry.Do(func() error {
-		srcUpdateHeader, dstUpdateHeader, err = GetIBCUpdateHeaders(ctx, srch, dsth, c.ChainProvider, dst.ChainProvider, c.ClientID(), dst.ClientID())
-		if err != nil {
+	var srcHeader, dstTrustedHeader provider.IBCHeader
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return retry.Do(func() error {
+			var err error
+			srcHeader, err = src.ChainProvider.QueryIBCHeader(egCtx, srch)
 			return err
-		}
-		return nil
+		}, retry.Context(ctx), RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
```

Comment:
I think, if the main function passed to retry.Do is using `egCtx`, the `retry.Context` call should also use `egCtx`: if one `eg.Go` fails, the other `retry.Do` will be canceled, but the outer context will continue retrying until the attempts exhaust.

If my reasoning is wrong, a clarifying comment would help.

Reply:
Yes you are correct, that would be better behavior. updated